src/core_init_atmosphere/mpas_parse_geoindex.F

```diff
@@ -0,0 +1,238 @@
+   use mpas_derived_types, only : MPAS_LOG_ERR, MPAS_LOG_WARN
+   use mpas_pool_routines
+
+   implicit none
+
+   private
+
+   public :: mpas_parse_index
+
+   contains
+
+   !***********************************************************************
+   !
+   !  subroutine mpas_parse_geoindex
+   !
+   !> \brief   Parse a geogrid's index file and put the results into an MPAS pool
+   !> \author  Miles A. Curry 
+   !> \date    02/20/2020
+   !> \details 
+   !> Parse an index file of a static data set into an MPAS pool, allocating
+   !> each keyword=value pair into the pool with the pool member key being 
+   !> keyword, and the value being value.
+   !>
+   !> This function can parse index files with one keyword=value pair 
+   !> per line; a "#" at the start of a line, which will cause the line to be 
+   !> ignored; or an empty line containing only a newline/return character, which
+   !> will also be ignored. Spaces or tabs before, between or after the 
+   !> keyword=value tokens are > ignored.
+   !>
+   !> If a line contains anything but the above valid syntaxes, a syntax
+   !> error will be raised and the model will fail.
+   !>
+   !> Case is ignored.
+   !>
+   !> The definitions of a keyword, which can found in section 3-53
+   !> of the WRF-AWR User's Guide, will determine the corresponding type
+   !> of that keyword. A keyword that has been assigned the wrong type
+   !> will raise a type error which cause the model to fail.
+   !>
+   !> Keywords that are not handled explicitly by this function will produce
+   !> a warning that the keyword is unrecognized.
+   !
+   !-----------------------------------------------------------------------
+   subroutine mpas_parse_index(path, geo_pool)
+
+      implicit none
+      ! Input Variables
+      character (len=StrKIND), intent(in) :: path
+      type (mpas_pool_type), intent(inout), pointer :: geo_pool
+
+      ! Local Variables
+      character (len=StrKIND) :: index_path, line, lhs, rhs
+      character (len=StrKIND) :: read_err_msg, open_msg
+      integer :: open_stat, read_stat, line_read_stat
+      integer :: i, k
+      logical :: res
+
+      character (len=StrKIND), pointer :: char_t
+      integer :: iceiling, ifloor
+      integer, pointer :: int_t
+      real, pointer :: real_t
+
+      index_path = trim(path)//"index"
+      inquire(file=index_path, exist=res)
+
+      if ( .not. res) then
+         call mpas_log_write("No 'index' file found in the geographic data set: "//index_path, messageType=MPAS_LOG_CRIT)
+      endif
+
+      open_stat = 0
+
+      open(10, FILE=index_path, action='READ', iostat=open_stat, iomsg=open_msg)
+      if (open_stat /= 0) then
+         call mpas_log_write("Could not open 'index' file at:'"//trim(index_path)//"'", messageType=MPAS_LOG_ERR)
+         call mpas_log_write(trim(open_msg), messageType=MPAS_LOG_CRIT)
+      endif
+
+      line_read_stat = 0
+      read_stat = 0
+      k = 0
+      do while ( .true. )
```

Comment:
While there are good reasons for using a `while (.true.)` loop in some situations, I think it's harder for the reader to immediately appreciate the conditions under which the loop will terminate. I think it's worth considering a loop structure like:
```fortran
      k = 1
      read(10,'(A)', iostat=line_read_stat) line
      do while (line_read_stat == 0)

         ... work ...

         k = k + 1 ! Keep track of line numbers for error reporting
         read(10,'(A)', iostat=line_read_stat) line
      end do
```

Reply:
I thought about doing this method, but I wasn't sure which method would be most readable. I think the version you suggested is most readable; however, I'm not sure that it will work with the current implementation.

On lines 94 and 95, we check to see if the line is comment (Starts with '#') or an empty line. If it is, `cycle` is called, the loops starts over, execution returns to line 94 and 95 with the same value for line, and again calls cycle. Thus the program enters an infinite loop.

https://github.com/MPAS-Dev/MPAS-Model/blob/b262a9cf4a720130823a1dee6338da6dd437b72d/src/core_init_atmosphere/mpas_parse_geoindex.F#L94-L95

If we move the cycle lines to below the read at the end of the `do while` loop, then we potentially could try to parse a comment or an empty line that was read by the first read.

To fix it, we could wrap the `do` loop on line 97 in an `if` statement that checks for either a comment or an empty line and avoid using the `cycle` statement all together.

Do you have any thoughts on this @mgduda?

Reply:
Would it work to replace the two lines (94 and 95) that handle comment and blank lines with something like the following?
```fortran
         !
         ! If a blank or comment line is encountered, just move on to the next line
         !
         if (line(1:1) == '#' .or. len_trim(line) == 0) then
            k = k + 1
            read(10,'(A)', iostat=line_read_stat) line
            cycle
         end if
```